src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/ConstantExpectedAnalyzer.cs

```diff
@@ -0,0 +1,566 @@
+                Id,
+                s_localizableUsageTitle,
+                CreateLocalizableResourceString(nameof(ConstantExpectedAttributExpectedMessage)),
+                DiagnosticCategory.Performance,
+                Level,
+                description: s_localizableUsageDescription,
+                isPortedFxCopRule: false,
+                isDataflowRule: false);
+        }
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+            CA1860.UnsupportedTypeRule, CA1860.IncompatibleConstantTypeRule,
+            CA1860.InvalidBoundsRule, CA1860.InvertedRangeRule,
+            CA1861.ConstantOutOfBoundsRule, CA1861.ConstantInvalidConstantRule,
+            CA1861.ConstantNotConstantRule, CA1861.AttributeExpectedRule);
+
+        protected abstract DiagnosticHelper Helper { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(context => OnCompilationStart(context));
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            context.RegisterOperationAction(OnInvocation, OperationKind.Invocation);
+            context.RegisterSymbolAction(context => OnMethodSymbol(context), SymbolKind.Method);
+            RegisterAttributeSyntax(context);
+            return;
+        }
+
+        private static void OnMethodSymbol(SymbolAnalysisContext context)
+        {
+            var methodSymbol = (IMethodSymbol)context.Symbol;
+            if (TryGetMethodInterface(methodSymbol, out var interfaceMethodSymbol))
+            {
+                CheckAttribute(context, methodSymbol.Parameters, interfaceMethodSymbol.Parameters);
+            }
+            else if (methodSymbol.OverriddenMethod is not null)
+            {
+                CheckAttribute(context, methodSymbol.Parameters, methodSymbol.OverriddenMethod.Parameters);
+            }
+
+            static void CheckAttribute(SymbolAnalysisContext context, ImmutableArray<IParameterSymbol> parameters, ImmutableArray<IParameterSymbol> baseParameters)
+            {
+                for (var i = 0; i < parameters.Length; i++)
+                {
+                    var parameter = parameters[i];
+                    if (!IsConstantCompatible(parameter.Type))
+                    {
+                        continue;
+                    }
+                    var baseParameter = baseParameters[i];
+                    if (HasConstantExpectedAttributeData(baseParameter) && !HasConstantExpectedAttributeData(parameter))
+                    {
+                        // mark the parameter including the type and name
+                        var diagnostic = parameter.DeclaringSyntaxReferences[0].GetSyntax().CreateDiagnostic(CA1861.AttributeExpectedRule);
```

Comment:
In any case, is it guaranteed that the index access `parameter.DeclaringSyntaxReferences[0]` does not throw an `IndexOutOfRangeException`?

Reply:
I believe so, we are checking the method declaration here. So if we have the symbol already, the syntax has to exist and be valid